route-rs-runtime/src/packet/ipv6.rs

```diff
@@ -0,0 +1,336 @@
+        // overrun our array trying to access the entire payload.
+        let payload_len = u16::from_be_bytes([packet[14 + 4], packet[14 + 5]]) as usize;
+        let packet_len = packet.len();
+        if payload_len + 14 + 40 > packet_len {
+            return Err("Packet has invalid payload len field");
+        }
+
+        Ok(Ipv6Packet {
+            data: packet,
+            packet_offset: 14,
+            payload_offset: packet_len - payload_len,
+        })
+    }
+
+    pub fn traffic_class(&self) -> u8 {
+        ((self.data[self.packet_offset] & 0x0F) << 4)
+            + (self.data[self.packet_offset + 1] & 0xF0 >> 4)
+    }
+
+    pub fn flow_label(&self) -> u32 {
+        u32::from_be_bytes([
+            0,
+            self.data[self.packet_offset + 1] & 0x0F,
+            self.data[self.packet_offset + 2],
+            self.data[self.packet_offset + 3],
+        ])
+    }
+
+    pub fn payload_length(&self) -> u16 {
+        u16::from_be_bytes([
+            self.data[self.packet_offset + 4],
+            self.data[self.packet_offset + 5],
+        ])
+    }
+
+    pub fn next_header(&self) -> IpProtocol {
+        IpProtocol::from(self.data[self.packet_offset + 6])
+    }
+
+    pub fn set_next_header(&mut self, header: u8) {
+        self.data[self.packet_offset + 6] = header;
+    }
+
+    pub fn hop_limit(&self) -> u8 {
+        self.data[self.packet_offset + 7]
+    }
+
+    // Is there a bug here if there is no payload? wonder if
+    // self.payload_offset would overrun the array, and cause a panic
+    pub fn payload(&self) -> Cow<[u8]> {
+        Cow::from(&self.data[self.payload_offset..])
+    }
+
+    pub fn set_payload(&mut self, payload: &[u8]) {
+        let new_payload_len = payload.len();
+        self.data.truncate(self.payload_offset);
+
+        let payload_len = (new_payload_len as u16).to_be_bytes();
+        self.data[self.packet_offset + 4..self.packet_offset + 6].copy_from_slice(&payload_len);
+
+        self.data.reserve_exact(new_payload_len);
+        self.data.extend(payload);
+        self.payload_offset = self.data.len() - payload.len();
+    }
+
+    pub fn src_addr(&self) -> Ipv6Addr {
+        Ipv6Addr::from_byte_slice(&self.data[self.packet_offset + 8..self.packet_offset + 24])
+            .unwrap()
+    }
+
+    pub fn dest_addr(&self) -> Ipv6Addr {
+        Ipv6Addr::from_byte_slice(&self.data[self.packet_offset + 24..self.packet_offset + 40])
+            .unwrap()
+    }
+
+    pub fn set_src_addr(&mut self, addr: Ipv6Addr) {
+        self.data[self.packet_offset + 8..self.packet_offset + 24]
+            .copy_from_slice(&addr.bytes()[..]);
+    }
+
+    pub fn set_dest_addr(&mut self, addr: Ipv6Addr) {
+        self.data[self.packet_offset + 24..self.packet_offset + 40]
+            .copy_from_slice(&addr.bytes()[..]);
+    }
+
+    //TODO: Test the get and set for extension headers.
+    pub fn extension_headers(&self) -> Vec<Cow<[u8]>> {
```

Comment:
Use a common function (between this and `get_ipv6_payload_type`)

Reply:
Took at look at this, while a good amount of the logic is the same. The main loop differs which makes abstracting the logic hard.